### DISCUSSION TITLE
Sanic integration - `on_request`

### DISCUSSION THREAD

**Discussion by lllama** [ahopkins/mayim#39] [Q&A]
Is there a way to have the executors injected in the `on_request` handlers? Would be useful for session handling.

**Comment by lllama** [comment]
Just realised I can use the `executor = Mayim.get(MyExecutor)` pattern. But injecting as a parameter feels more consistent.

**Comment by ahopkins** [comment]
You should be able to access them as injectables:

```python
from typing import List
from dataclasses import asdict
from sanic import Sanic, Request, json
from sanic_ext import Extend
from mayim import Mayim
from mayim.executor import Executor
from mayim.extensions import SanicMayimExtension


class CityExecutor(Executor):
    async def select_all_cities(
        self, limit: int = 4, offset: int = 0
    ) -> List[City]:
        ...


app = Sanic(__name__)
Extend.register(
    SanicMayimExtension(
        executors=[CityExecutor], dsn="postgres://..."
    )
)


@app.get("/")
async def handler(request: Request, executor: CityExecutor):
    cities = await executor.select_all_cities()
    return json({"cities": [asdict(city) for city in cities]})
```
[source](https://ahopkins.github.io/mayim/guide/extensions.html#sanic)

Is this what you meant?

**Reply by lllama** [reply]
I'm doing some session checking with an `on_request` handler. So:

```py
@app.on_request
async def handler(request: Request, executor: CityExecutor):
    cities = await executor.select_all_cities()
    return json({"cities": [asdict(city) for city in cities]})
```

but the executor doesn't get injected. I can do this though:
```py
@app.on_request
async def handler(request: Request):
    executor = Mayim.get(CityExecutor)
    cities = await executor.select_all_cities()
    return json({"cities": [asdict(city) for city in cities]})
```

**Reply by ahopkins** [reply]
Ahh... I see. :thinking: 

I think this is not a Mayim issue but more a limitation with Sanic Extensions. I am fairly certain we never added injection into middleware. Although, it does sound like a reasonable request. 

I'd open it as [an issue there](https://github.com/sanic-org/sanic-ext/issues).
